v2/dtos/requests/device.go

```diff
@@ -0,0 +1,167 @@
+	}
+	if err := json.Unmarshal(b, &alias); err != nil {
+		return v2.NewErrContractInvalid("Failed to unmarshal request body as JSON.")
+	}
+
+	*d = AddDeviceRequest(alias)
+
+	// validate AddDeviceRequest DTO
+	if err := d.Validate(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// AutoEventDTOsToModels transforms the AutoEvent DTO array to the AutoEvent model array
+func AutoEventDTOsToModels(autoEventDTOs []dtos.AutoEvent) []models.AutoEvent {
+	autoEvents := make([]models.AutoEvent, len(autoEventDTOs))
+	for i, a := range autoEventDTOs {
+		autoEvents[i] = dtos.ToAutoEventModel(a)
+	}
+	return autoEvents
+}
+
+// ProtocolDTOsToModels transforms the Protocol DTO map to the Protocol model map
+func ProtocolDTOsToModels(protocolDTOs map[string]dtos.ProtocolProperties) map[string]models.ProtocolProperties {
+	protocols := make(map[string]models.ProtocolProperties)
+	for k, protocolProperties := range protocolDTOs {
+		protocols[k] = dtos.ToProtocolPropertiesModel(protocolProperties)
+	}
+	return protocols
+}
+
+// AddDeviceReqToDeviceModels transforms the AddDeviceRequest DTO array to the Device model array
+func AddDeviceReqToDeviceModels(addRequests []AddDeviceRequest) (Devices []models.Device) {
+	for _, req := range addRequests {
+		var d models.Device
+		d.Name = req.DeviceName
+		d.ServiceName = req.ServiceName
+		d.ProfileName = req.ProfileName
+		d.AdminState = models.AdminState(req.AdminState)
+		d.OperatingState = models.OperatingState(req.OperatingState)
+		d.AutoEvents = AutoEventDTOsToModels(req.AutoEvents)
+		d.Protocols = ProtocolDTOsToModels(req.Protocols)
+		Devices = append(Devices, d)
+	}
+	return Devices
+}
+
+// UpdateDeviceRequest defines the Request Content for PUT event as pushed DTO.
+// This object and its properties correspond to the UpdateDeviceRequest object in the APIv2 specification:
+// https://app.swaggerhub.com/apis-docs/EdgeXFoundry1/core-metadata/2.x#/UpdateDeviceRequest
+type UpdateDeviceRequest struct {
+	common.BaseRequest `json:",inline"`
+	Id                 string                             `json:"Id" validate:"required"`
+	DeviceName         string                             `json:"deviceName" validate:"required"`
+	ServiceName        string                             `json:"serviceName"`
+	ProfileName        string                             `json:"profileName"`
+	AdminState         string                             `json:"adminState"`
+	OperatingState     string                             `json:"operatingState"`
+	AutoEvents         []dtos.AutoEvent                   `json:"autoEvents"`
+	Protocols          map[string]dtos.ProtocolProperties `json:"protocols"`
+	LastConnected      int64                              `json:"lastConnected"`
+	LastReported       int64                              `json:"lastReported"`
+	Notify             bool                               `json:"notify"`
+}
+
+// Validate satisfies the Validator interface
+func (dp UpdateDeviceRequest) Validate() error {
```

Comment:
add logic to validate AutoEvent.Frequency format

Reply:
Added.